### ISSUE TITLE
`d_array.map` only takes endofunctions

### ISSUE THREAD

**Issue by jcommelin** [leanprover-community/lean#184]
```lean
#check @d_array.map
d_array.map : Π {n : ℕ} {α : fin n → Type u_1}, (Π (i : fin n), α i → α i) → d_array n α → d_array n α
```
I think this should also take a `\beta` and maps from `\alpha i` to `\beta i` instead of only `\alpha i → \alpha i`.

### Prerequisites

* [ ] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover-community/lean/issues).
        * Specifically, check out the [wishlist](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AI-wishlist), open [RFCs](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3ARFC),
        or [feature requests](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AFeature).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

[Description of the issue]

### Steps to Reproduce

1. [First Step]
2. [Second Step]
3. [and so on...]

**Expected behavior:** [What you expect to happen]

**Actual behavior:** [What actually happens]

**Reproduces how often:** [What percentage of the time does it reproduce?]

### Versions

You can get this information from copy and pasting the output of `lean --version`,
please include the OS and what version of the OS you're running.

### Additional Information

Any additional information, configuration or data that might be necessary to reproduce the issue.

**Comment by gebner** [comment]
Do you want to submit a PR for this?

**Comment by digama0** [comment]
This is deliberate. Probably the name of the function should change, but the endofunction `map` has a modify in place implementation that general map cannot.

**Comment by gebner** [comment]
Can you explain why a type-changing `map` couldn't have an in-place implementation?  It works in Lean 4: https://github.com/leanprover/lean4/blob/7322345768e59dd89a85c9e6a635372972990f06/src/Init/Data/Array/Basic.lean#L498-L499

**Comment by digama0** [comment]
That doesn't look like an in place implementation to me, it's just pushing to a new array

**Comment by digama0** [comment]
If there was a type changing `d_array.write` primitive, it might be possible to write a type changing in place map, but in the intermediate state you would have an array where the first half has type `A i` and the second half has type `B i`

**Comment by gebner** [comment]
Of course, you can't do `d_array.write`.  But I don't see anything wrong with `d_array.map`

**Comment by digama0** [comment]
I'm saying that the current implementation of `d_array.map` is type constrained exactly because it iterates over the array replacing `a[i] := f(a[i])`, and as a result it requires no additional memory. There is an alternative implementation of `map` that constructs a new array, and this can easily be type changing. (It's trivial to write this implemenation using the constructor and function composition.) But it will require a temporary.

I propose we have both functions, renaming the existing `map` function to `endomap` or something.

**Comment by gebner** [comment]
Mario, the arrays in Lean are untyped.  They are essentially an `std::vector<vm_obj>` (and after #182 also literally).  There are no specialized integer arrays or character arrays that could cause an issue here.

Thanks for reminding me of the `{rev_,}iterate_aux` functions.  They are completely pointless since the VM overrides the `iterate` function anyhow.  I will remove them.

**Comment by digama0** [comment]
I'm aware of this, I am talking about lean typing. The fact that they are C++-untyped means that a type changing `d_array.write` primitive is actually possible, and could be used to implement a map operation where the type gradually changes from `array n A` to `array n B` via `d_array n (\i, if i < n then A else B)`.

**Comment by digama0** [comment]
If you want to have a C++ implementation of `map`, then there are no problems; you can just implement it in place even though the intermediate values aren't well typed, and use the naive implementation for the lean model. But I would prefer to expose the mechanisms necessary to efficiently implement this in lean rather than hard-code everything.

**Comment by gebner** [comment]
The C++ implementation for `map` is already there (it's called `foreach` for some reason).  I just want to give it a better type.  If you want to contribute a type-changing `write` function, that's also ok.

**Comment by gebner** [comment]
Fixed in #190.
